### PR TITLE
[NVFP4 MoE] Skip dead blockscale_swizzled allocation on Marlin fallback (SM80-SM90x), saves ~3.5 GiB/GPU

### LINKED ISSUE #38073: [NVFP4 MoE Marlin fallback] w13/w2_blockscale_swizzled dead copies waste ~3.5 GiB per GPU (SM80–SM90x)
**Environment**: 2× NVIDIA L20 (sm_89, 48GB each), image `lmsysorg/sglang:qwen38flashnext` (sglang 0.0.0.dev1+git593134d17, torch 2.13.0+cu130), model Qwen3.8-Flash-Next (qwen4_exp: 48 layers, 512 routed experts) with a RadixArk NVFP4 W4A4 gs16 checkpoint, TP2, default `auto` MoE backend → Marlin fallback.

The same code exists on `main` (see `modelopt_quant.py` ~L2404-2427) and on branch `qwen4-main-squashed` @ 78c5024.

## Summary

When the NVFP4 fused-MoE path falls back to Marlin (`use_marlin_fallback = (8,0) <= capability < (10,0)`), `create_weights` still allocates full-size `w13_blockscale_swizzled` / `w2_blockscale_swizzled` Parameters. On this path they are:

1. built from the **still-uninitialized** `torch.empty` scale buffer (they permanently hold garbage bytes),
2. never re-swizzled after checkpoint load,
3. never read by any kernel at runtime.

For our 48-layer / 512-expert model this pins **~3.5 GiB of dead memory per rank (TP2)** for the whole server lifetime. On 48GB-class cards this is the difference between fitting and OOM.

## Root cause walk-through

1. `ModelOptNvFp4FusedMoEMethod.create_weights` (~L2405-2430 on the branch): the swizzled copies are created whenever `enable_flashinfer_trtllm_moe` is false — at which point `w13_weight_scale` is still `torch.empty(...)`:
```python
layer.w13_blockscale_swizzled = Parameter(
    swizzle_blockscale(layer.w13_weight_scale), requires_grad=False
)
```
2. `process_weights_after_loading`: the Marlin branch (~L2496-2541) calls `prepare_moe_nvfp4_layer_for_marlin(layer)` and **returns early**. The CuteDSL branch (which re-swizzles with real data, ~L2741+) and the TRT-LLM branch (which aliases the placeholder precisely to free it, ~L2705-2711: *"TRTLLM doesn't read *_blockscale_swizzled; alias to free the placeholders from create_weights"*) are never reached.
3. `prepare_moe_nvfp4_layer_for_marlin` (`marlin_utils_fp4.py`) replaces `w13/w2_weight`, `w13/w2_weight_scale`, `*_scale_2`, workspace — but never touches `blockscale_swizzled`. (The MXFP4-Marlin sibling path does `delattr` its loader-format leftovers; the NVFP4-Marlin path misses this cleanup.)
4. The Marlin MoE kernel consumes only the processed `w13/w2_weight_scale`. Nothing on this path ever reads `*_blockscale_swizzled`.

## Size math and measurement

Packed expert weights 28.1 GiB/rank ⇒ 56.2G fp4 elements ⇒ gs16 scales = 3.51G × 1 byte ⇒ **3.51 GiB of dead swizzled copies per rank** (w13+w2 across 48 layers).

Measured with the fix below: `Load weight end` memory **41.24 → 37.68 GiB** (−3.56 GiB/rank). Inference outputs unchanged (greedy needle-in-haystack at 249K context, vision, and code-gen regression suite all identical), as expected — the removed tensors had no reader.

## Proposed fix

Mirror the existing TRT-LLM None-placeholder pattern:

```python
# __init__: keep the flag
self.use_marlin_fallback = use_marlin_fallback

# create_weights:
if self.enable_flashinfer_trtllm_moe or self.use_marlin_fallback:
    layer.w13_blockscale_swizzled = None
else:
    layer.w13_blockscale_swizzled = Parameter(swizzle_blockscale(...), ...)
# (same for w2)
```

Happy to open a PR with this if you'd like.

## Impact

All SM80–SM90x GPUs serving NVFP4 W4A4 MoE checkpoints with the default `auto` backend (A100/H100/L40/L20/3090/4090…). The waste scales with `num_experts × intermediate_size × hidden_size / group_size` — worst exactly on the large-MoE + small-VRAM combinations the Marlin fallback exists for.

### PR DESCRIPTION
**Pull request by JUNQINGV587**
## Summary

On the NVFP4 fused-MoE **Marlin fallback path** (SM80–SM90x, i.e. `use_marlin_fallback = (8,0) <= capability < (10,0)`), `create_weights` allocates full-size `w13_blockscale_swizzled` / `w2_blockscale_swizzled` Parameters that are:

1. built from the still-uninitialized `torch.empty` scale buffer (they hold garbage bytes forever),
2. never re-swizzled after checkpoint load (`process_weights_after_loading` returns early in the Marlin branch, before the CuteDSL re-swizzle / TRT-LLM alias),
3. never read by any kernel at runtime (`prepare_moe_nvfp4_layer_for_marlin` + the Marlin kernel consume only the processed `w13/w2_weight_scale`; unlike the MXFP4-Marlin sibling path, no `delattr` cleanup happens here).

On a 48-layer / 512-expert NVFP4 model (Qwen3.8-Flash-Next, RadixArk W4A4 gs16 checkpoint) at TP2, these dead copies pin **~3.5 GiB per rank** for the server's whole lifetime. On 48GB-class cards this alone can decide fit-vs-OOM.

## Fix

Skip the allocation on the Marlin fallback path, mirroring the existing TRT-LLM `None`-placeholder pattern (the flag computed in `__init__` is persisted as `self.use_marlin_fallback`). 10-line diff, zero numerical change — the removed tensors had no reader.

## Measurement (2× NVIDIA L20, sm_89, TP2)

| | before | after |
|---|---|---|
| `Load weight end` mem usage / rank | 41.24 GiB | **37.68 GiB** (−3.56) |
| regression suite | — | 262K-context needle, vision, code-gen all pass with byte-identical outputs |

Full root-cause walk-through with file:line references in the linked issue.

Closes #38073

---
### CI States

Latest PR Test (Base): :x: [Run #33955099356](https://github.com/sgl-project/sglang/actions/runs/33955099356)
Latest PR Test (Extra): :x: [Run #33955099129](https://github.com/sgl-project/sglang/actions/runs/33955099129)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33955099291](https://github.com/sgl-project/sglang/actions/runs/33955099291)

### CODE REVIEW

**Comment by JUNQINGV587** [comment]
Production measurements on 2×L20 (sm_89) TP2, Qwen3.8-Flash-Next-NVFP4 (RadixArk checkpoint, marlin fallback):

- Before: `torch.cuda.memory_allocated()` grows ~+0.66 GiB per layer during repack; 41.24 GiB allocated per rank after 48 layers.
- After this patch: flat 37.68 GiB per rank (≈3.5 GiB/rank reclaimed, matches one layer's packed weights + fp8 scales).

This is currently load-bearing for 48GB Ada cards: without it the NVFP4 MoE marlin path OOMs around layer 9/48 (also reported independently in #38074 by a third party on the identical stack).

CI is stuck at pr-gate (external PR missing the `run-ci` label — failures are not code-related). Could a maintainer add the label? @AniZpZ
